arbor/morph/locset.cpp

```diff
@@ -0,0 +1,248 @@
+#include <algorithm>
+#include <iostream>
+#include <numeric>
+
+#include <arbor/morph/error.hpp>
+#include <arbor/morph/locset.hpp>
+#include <arbor/morph/region.hpp>
+
+#include "util/rangeutil.hpp"
+#include "util/strprintf.hpp"
+
+#include "morph/em_morphology.hpp"
+
+namespace arb {
+namespace ls {
+
+//
+// Functions for taking the sum, union and intersection of location_lists (multisets).
+//
+//
+
+using it_t = mlocation_list::iterator;
+using const_it_t = mlocation_list::const_iterator;
+
+// Advance an iterator to the first value that is not equal to its current
+// value, or end, whichever comes first.
+template <typename T>
+T next_unique(T& it, T end) {
+    const auto& x = *it;
+    ++it;
+    while (it!=end && *it==x) ++it;
+    return it;
+};
+
+// Return the number of times that the value at it is repeated. Advances the
+// iterator to the first value not equal to its current value, or end,
+// whichever comse first.
+template <typename T>
+int multiplicity(T& it, T end) {
+    const auto b = it;
+    return std::distance(b, next_unique(it, end));
+};
+
+mlocation_list sum(const mlocation_list& lhs, const mlocation_list& rhs) {
+    mlocation_list v;
+    v.resize(lhs.size() + rhs.size());
+    std::merge(lhs.begin(), lhs.end(), rhs.begin(), rhs.end(), v.begin());
+    return v;
+}
+
+mlocation_list join(const mlocation_list& lhs, const mlocation_list& rhs) {
+    mlocation_list L;
+    L.reserve(lhs.size()+rhs.size());
+
+    auto l    = lhs.begin();
+    auto lend = lhs.end();
+    auto r    = rhs.begin();
+    auto rend = rhs.end();
+
+    auto at_end = [&]() { return l==lend || r==rend; };
+    while (!at_end()) {
+        auto x = (*l<*r) ? *l: *r;
+        auto count = (*l<*r)? multiplicity(l, lend):
+                     (*r<*l)? multiplicity(r, rend):
+                     std::max(multiplicity(l, lend), multiplicity(r, rend));
+        while (count--) L.push_back(x);
```

Comment:
Here and below (line 88), an alternative is to use `L.insert(L.end(), count, x)`. I'm sure it makes negligible performance difference.

Reply:
Done.